### PR TITLE
15749 addressing customer feedback

### PR DESCRIPTION
**Pull request by tyao117**
### Description

This PR address these issues raised by a customer: 
- I suggest to rephrase the info message along the lines of: “The name to give the root key that will be created to encrypt your stored secrets. Not used if supplying an existing key.”
- I suggest to rename the default value: sm-cos-key → sm-data-encryption-key
- Its not clear what is the propose of the skip_kms_iam_authorization_policy field.
- I suggest to rename the default value: sm-cos-key-ring → sm-data-encryption-key-ring
- Only used if provision_sm_instance is set to true” ? The context of this DA is to provision an SM instance, also there is no provision_sm_instance field option in this page.
- Its not clear what is the propose of the skip_en_iam_authorization_policy field.
- Given the required name of the resource group to provision the instance of SM, what is the perpose of the optional existing_resource_group field? 

Related to: https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/issues/79

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Review by ocofaigh — COMMENTED** [review]
@tyao117 In every single other module and DA, we use the terminology `skip_...` to handle IAM auth policy creation. By changing it to `create_...` its no longer consistent with anything else. What was the reasoning behind this?

**Comment by ocofaigh** [comment]
Also since you are updating var names here, you will need to update the tests. They currently fail with:
```
 2024/03/28 18:55:42 Terraform plan | Error: Invalid function argument
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan |   on ../../main.tf line 15, in locals:
 2024/03/28 18:55:42 Terraform plan |   15:   validate_auth_policy = var.kms_encryption_enabled && var.skip_kms_iam_authorization_policy == false && var.existing_kms_instance_guid == null ? tobool("When var.skip_kms_iam_authorization_policy is set to false, and var.kms_encryption_enabled to true, a value must be passed for var.existing_kms_instance_guid in order to create the auth policy.") : true
 2024/03/28 18:55:42 Terraform plan |     ├────────────────
 2024/03/28 18:55:42 Terraform plan |     │ while calling tobool(v)
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan | Invalid value for "v" parameter: cannot convert "When
 2024/03/28 18:55:42 Terraform plan | var.skip_kms_iam_authorization_policy is set to false, and
 2024/03/28 18:55:42 Terraform plan | var.kms_encryption_enabled to true, a value must be passed for
 2024/03/28 18:55:42 Terraform plan | var.existing_kms_instance_guid in order to create the auth policy." to bool;
 2024/03/28 18:55:42 Terraform plan | only the strings "true" or "false" are allowed.
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan | Error: Missing required argument
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan |   with module.kms[0].module.kms_key_rings["sm-data-encryption-key-ring"].ibm_kms_key_rings.key_ring,
 2024/03/28 18:55:42 Terraform plan |   on .terraform/modules/kms.kms_key_rings/main.tf line 7, in resource "ibm_kms_key_rings" "key_ring":
 2024/03/28 18:55:42 Terraform plan |    7:   instance_id   = var.instance_id
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan | The argument "instance_id" is required, but no definition was found.
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan | Error: Error in function call
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan |   on .terraform/modules/kms/main.tf line 28, in locals:
 2024/03/28 18:55:42 Terraform plan |   28:   existing_instance_validate_check = regex("^${local.existing_instance_validate_msg}$", (!local.existing_instance_validate_condition ? local.existing_instance_validate_msg : ""))
 2024/03/28 18:55:42 Terraform plan |     ├────────────────
 2024/03/28 18:55:42 Terraform plan |     │ while calling regex(pattern, string)
 2024/03/28 18:55:42 Terraform plan |     │ local.existing_instance_validate_condition is true
 2024/03/28 18:55:42 Terraform plan |     │ local.existing_instance_validate_msg is "A value must be provided for 'existing_key_protect_instance_guid' when 'create_key_protect_instance' is false"
 2024/03/28 18:55:42 Terraform plan | 
 2024/03/28 18:55:42 Terraform plan | Call to function "regex" failed: pattern did not match any part of the given
 2024/03/28 18:55:42 Terraform plan | string.
 2024/03/28 18:55:42 [1m[31mTerraform PLAN error: Terraform PLAN errorexit status 1[39m[0m
```

**Comment by tyao117** [comment]
Making the changes to the code to address the changes that are in the box note.

File: solutions/standard/variables.tf [hunk 1]
```diff
@@ -124,6 +118,6 @@ variable "existing_en_instance_crn" {
 
 variable "skip_en_iam_authorization_policy" {
   type        = bool
-  description = "Set to true to skip the creation of an IAM authorization policy that permits all Secrets Manager instances (scoped to the resource group) an 'Event Source Manager' role to the given Event Notifications instance passed in the `existing_en_instance_crn` input variable."
-  default     = false
+  description = "A flag to create an IAM authorization that grants Secrets Manager access to a Event Notification resource(true) or not(false). Set to false if a preexisting authorization can be used."
+  default     = true
```
Comment (ocofaigh) [inline solutions/standard/variables.tf:122]: by default we want to create the auth policy, so change this back to false

File: solutions/standard/variables.tf [hunk 2]
```diff
@@ -68,26 +68,20 @@ variable "secret_manager_tags" {
 
 variable "skip_kms_iam_authorization_policy" {
   type        = bool
-  description = "Set to true to skip the creation of an IAM authorization policy that permits all Secrets Manager instances in the resource group to read the encryption key from the KMS instance. If set to false, pass in a value for the KMS instance in the existing_kms_instance_guid variable."
-  default     = false
+  description = "A flag to create an IAM authorization that grants Secrets Manager access to a Key Protect resource(true) or not(false). Set to false if a preexisting authorization can be used."
```
Comment (ocofaigh) [inline solutions/standard/variables.tf:71]: description should match whatever format we go with for `skip_en_iam_authorization_policy`

File: ibm_catalog.json [hunk 3]
```diff
@@ -78,6 +78,67 @@
...
+                  "default_value": "base-security-services-cos",
+                  "description": "The name to use when creating the Cloud Object Storage instance.",
+                  "required": false
+              },
+              {
+                "key": "region",
+                "type": "string",
+                "required": true,
+                "default_value": "",
+                "options": [
```
Comment (ocofaigh) [inline ibm_catalog.json:95]: This list should probably be in alphabetical order as I think its an ordered list. Its also missing these supported regions:

- Sao Paulo
- Toronto
- BNPP_EU (do we want to include this or not?)
- Washington DC

File: solutions/standard/variables.tf [hunk 4]
```diff
@@ -124,6 +118,6 @@ variable "existing_en_instance_crn" {
 
 variable "skip_en_iam_authorization_policy" {
   type        = bool
-  description = "Set to true to skip the creation of an IAM authorization policy that permits all Secrets Manager instances (scoped to the resource group) an 'Event Source Manager' role to the given Event Notifications instance passed in the `existing_en_instance_crn` input variable."
-  default     = false
+  description = "A flag to create an IAM authorization that grants Secrets Manager access to a Event Notification resource(true) or not(false). Set to false if a preexisting authorization can be used."
```
[the remark this answers is not in the document]
Comment (SirSpidey) [reply solutions/standard/variables.tf:121]: > Its not clear what is the propose of the skip_kms_iam_authorization_policy field.

### original
> Set to true to skip the creation of an IAM authorization policy that permits all Secrets Manager instances (scoped to the resource group) an 'Event Source Manager' role to the given Event Notifications instance passed in the `existing_en_instance_crn` input variable.

### Update in PR
> A flag to create an IAM authorization that grants Secrets Manager access to a Event Notification resource(true) or not(false). Set to false if a preexisting authorization can be used.

Conall
> ...if the auth policy preexists, the value can be set to true to skip creation

---
### My suggestion

Whether to create an IAM authorization policy to permit sending notifications from Secrets Manager to the Event Notifications service that's specified in the `existing_en_instance_crn` input variable. Set to `false` to create the policy. Set to `true` if a policy with the correct access exists.

File: solutions/standard/variables.tf (hunk 2 above)
Comment (SirSpidey) [reply solutions/standard/variables.tf:71]: So if you go with my suggestion for `skip_en_iam_authorization_policy`,

Whether to create an IAM authorization policy to permit Secrets Manager instances to read the encryption key from the KMS instance. Set to `false` to create the policy, and specify the KMS instance in the `existing_kms_instance_guid` variable. Set to `true` if a policy with the correct access exists.

File: ibm_catalog.json [hunk 5]
```diff
@@ -78,6 +78,67 @@
...
                 "description": "This architecture supports creating and configuring IBM Secrets Manager instance."
               }
+            ],
+            "configuration": [
+              {
+                  "key": "cos_instance_name",
+                  "type": "string",
+                  "default_value": "base-security-services-cos",
+                  "description": "The name to use when creating the Cloud Object Storage instance.",
+                  "required": false
```
[the remark this answers is not in the document]
Comment (SirSpidey) [reply ibm_catalog.json:88]: Also. Always use either `IBM Cloud Object Storage` or `Object Storage`. Not Cloud Object Storage.

File: solutions/standard/variables.tf [hunk 6]
```diff
@@ -8,15 +8,15 @@ variable "ibmcloud_api_key" {
...
   type        = bool
-  description = "Whether to use an existing resource group."
+  description = "A flag to create a resource group(true) or to use a preexisting resource group (false) specified in the variable resource_group_name"
   default     = false
 }
 
 variable "resource_group_name" {
   type        = string
-  description = "The name of a new or an existing resource group in which to provision Secrets Manager resources to."
+  description = "The name of a new or an existing resource group in which to provision the Secrets Manager resources to."
```
Comment (SirSpidey) [inline solutions/standard/variables.tf:19]:
```suggestion
  description = "The name of a new or existing resource group where Secrets Manager instances are provisioned."
```

File: solutions/standard/variables.tf [hunk 7]
```diff
@@ -37,7 +37,7 @@ variable "secrets_manager_instance_name" {
 
 variable "service_plan" {
   type        = string
-  description = "The service/pricing plan to use when provisioning a new Secrets Manager instance. Allowed values: 'standard' and 'trial'. Only used if `provision_sm_instance` is set to true."
+  description = "The service/pricing plan to use when provisioning a new Secrets Manager instance. Allowed values: 'standard' and 'trial'."
```
Comment (SirSpidey) [inline solutions/standard/variables.tf:40]:
```suggestion
  description = "The pricing plan to use for new Secrets Manager instances. Allowed values: `standard`, `trial`."
```

File: solutions/standard/variables.tf [hunk 8]
```diff
@@ -68,26 +68,20 @@ variable "secret_manager_tags" {
...
-  description = "Set to true to skip the creation of an IAM authorization policy that permits all Secrets Manager instances in the resource group to read the encryption key from the KMS instance. If set to false, pass in a value for the KMS instance in the existing_kms_instance_guid variable."
-  default     = false
+  description = "A flag to create an IAM authorization that grants Secrets Manager access to a Key Protect resource(true) or not(false). Set to false if a preexisting authorization can be used."
+  default     = true
 }
 
 variable "existing_sm_kms_key_crn" {
   type        = string
-  description = "The CRN of an existing KMS key to use for Secrets Manager. If not supplied, a new key ring and key will be created."
+  description = "The CRN of an existing KMS key to use for Secrets Manager. If not supplied, a new key ring and new key will be created."
```
Comment (SirSpidey) [inline solutions/standard/variables.tf:77]: You don't need `new` in either object. You can't create something "old".

```suggestion
  description = "The CRN of a KMS key to use for Secrets Manager. If not specified, a key ring and key are created."
```

**Review by SirSpidey — COMMENTED** [review]
some suggestions

**Comment by tyao117** [comment]
closing this PR after @ocofaigh made a seperate PR to improve on this PR: https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/93

**Pull request closed without merging**